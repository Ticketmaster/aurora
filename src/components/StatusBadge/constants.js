import PropTypes from "prop-types";
import { themes } from "../../theme";

export const badgeProps = {
  label: PropTypes.string.isRequired,
  color: PropTypes.string
};

export const badgeDefault = {
  color: themes.global.primary.base
};
