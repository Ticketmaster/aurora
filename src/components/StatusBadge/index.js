export { default as StatusBadge } from "./StatusBadge";
export { default as StatusBadgeGroup } from "./StatusBadgeGroup";
