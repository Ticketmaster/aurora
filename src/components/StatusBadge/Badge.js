import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { typography, themes, constants } from "../../theme";

import { badgeTypes, badgeProps, badgeDefault } from "./constants";

export const StyledBadge = styled.div`
  padding: 4px 8px;
  font-weight: ${typography.weight.semiBold};
  text-align: center;
  color: ${({ type = "dark" }) =>
    type === "light" ? themes.global.gray01 : themes.global.white.base};
  font-size: ${typography.size.mini};
  border-radius: ${constants.borderRadius.small};
  letter-spacing: 0.25;
  text-transform: uppercase;
  background-color: ${({ color }) => color};
`;

const Badge = ({ type, color, label, className, ...rest }) => (
  <StyledBadge
    className={`${className} badge`}
    type={type}
    color={color}
    {...rest}
  >
    {label}
  </StyledBadge>
);

Badge.propTypes = {
  ...badgeProps,
  type: PropTypes.oneOf(Object.values(badgeTypes)),
  className: PropTypes.string
};

Badge.defaultProps = {
  ...badgeDefault,
  type: badgeTypes.dark,
  className: ""
};

Badge.displayName = "Badge";

export default Badge;
