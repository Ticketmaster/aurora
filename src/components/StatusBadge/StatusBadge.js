import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import classNames from "classnames";
import { typography, themes, constants, spacing } from "../../theme";

import { badgeProps, badgeDefault } from "./constants";
import { VARIANTS, DARK, LIGHT } from "../constants";

export const StyledBadge = styled.div`
  padding: ${spacing.slim} ${spacing.cozy};
  font-weight: ${typography.weight.semiBold};
  text-align: center;
  color: ${({ variant = DARK }) =>
    variant === LIGHT ? themes.global.gray01 : themes.global.white.base};
  font-size: ${typography.size.mini};
  line-height: 1.2;
  border-radius: ${constants.borderRadius.small};
  letter-spacing: 0.25;
  text-transform: uppercase;
  background-color: ${({ color }) => color};
`;

const StatusBadge = ({ variant, color, label, className, ...rest }) => (
  <StyledBadge
    className={classNames(className, `badge`)}
    variant={variant}
    color={color}
    {...rest}
  >
    {label}
  </StyledBadge>
);

StatusBadge.propTypes = {
  ...badgeProps,
  variant: PropTypes.oneOf(VARIANTS),
  className: PropTypes.string
};

StatusBadge.defaultProps = {
  ...badgeDefault,
  variant: DARK,
  className: ""
};

StatusBadge.displayName = "StatusBadge";

export default StatusBadge;
