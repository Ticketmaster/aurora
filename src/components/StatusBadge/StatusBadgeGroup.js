import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import StatusBadge, { StyledBadge } from "./StatusBadge";
import Tooltip from "../Tooltip";
import { themes, spacing } from "../../theme";

import { badgeProps } from "./constants";
import { VARIANTS, DARK, DIRECTIONS } from "../constants";

const StatusBadgeContainer = styled.div`
  display: flex;
  .badge {
    margin-left: ${spacing.cozy};

    &:first-child,
    &.hidden {
      margin-left: 0;
    }

    &.hidden {
      margin-top: ${spacing.cozy};
      :first-child {
        margin-top: 0;
      }
    }
  }
`;

const HiddenBadgesCount = styled(StyledBadge)`
  background: ${themes.global.transparent};
  color: ${themes.global.onyx.light};
`;

const BadgeTooltip = styled(Tooltip)`
  padding: 8px;
`;

class StatusBadgeGroup extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: false
    };

    this.mouseLeave = this.mouseLeave.bind(this);
    this.elementHovered = this.elementHovered.bind(this);
  }

  mouseLeave() {
    this.setState({
      isOpen: false
    });
  }

  elementHovered(e) {
    const data = Tooltip.getDimensionsFromEvent(e);
    this.setState({
      isOpen: true,
      ...data
    });
  }

  renderBadges = (badges, variant, additionalClass) =>
    badges.map(item => (
      <StatusBadge
        className={additionalClass}
        {...item}
        variant={variant}
        key={item.label}
      />
    ));

  render() {
    const {
      visibleBadges,
      hiddenBadges,
      tooltipDirection,
      variant
    } = this.props;
    const { isOpen, ...position } = this.state;

    return (
      <StatusBadgeContainer>
        {this.renderBadges(visibleBadges, variant)}
        {hiddenBadges &&
          hiddenBadges.length > 0 && (
            <HiddenBadgesCount
              onMouseEnter={this.elementHovered}
              onMouseLeave={this.mouseLeave}
            >
              {`+${hiddenBadges.length} more`}
            </HiddenBadgesCount>
          )}
        <BadgeTooltip
          direction={tooltipDirection}
          position={{ ...position }}
          isVisible={isOpen}
        >
          {this.renderBadges(hiddenBadges, variant, "hidden")}
        </BadgeTooltip>
      </StatusBadgeContainer>
    );
  }
}

StatusBadgeGroup.propTypes = {
  variant: PropTypes.oneOf(VARIANTS),
  visibleBadges: PropTypes.arrayOf(PropTypes.shape(badgeProps)),
  hiddenBadges: PropTypes.arrayOf(PropTypes.shape(badgeProps)),
  tooltipDirection: PropTypes.oneOf(DIRECTIONS)
};

StatusBadgeGroup.defaultProps = {
  variant: DARK,
  visibleBadges: [],
  hiddenBadges: [],
  tooltipDirection: "bottom"
};

StatusBadgeGroup.displayName = "StatusBadgeGroup";

export default StatusBadgeGroup;
