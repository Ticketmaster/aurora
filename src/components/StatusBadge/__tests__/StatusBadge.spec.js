import React from "react";
import renderer from "react-test-renderer";

import StatusBadge from "../StatusBadge";

describe("StatusBadge", () => {
  it("should match snapshot when variant is undefined", () => {
    const tree = renderer
      .create(<StatusBadge label="snapshot test" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
  it("should match snapshot when variant is dark", () => {
    const tree = renderer
      .create(<StatusBadge variant="dark" label="snapshot test" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
  it("should match snapshot when variant is light", () => {
    const tree = renderer
      .create(<StatusBadge variant="light" label="light test" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
