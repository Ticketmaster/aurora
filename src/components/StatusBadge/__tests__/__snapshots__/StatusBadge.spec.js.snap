// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBadge should match snapshot when variant is dark 1`] = `
<div
  className="badge gvbcbQ"
  color="#026cdf"
>
  snapshot test
</div>
`;

exports[`StatusBadge should match snapshot when variant is light 1`] = `
<div
  className="badge kxnhqq"
  color="#026cdf"
>
  light test
</div>
`;

exports[`StatusBadge should match snapshot when variant is undefined 1`] = `
<div
  className="badge gvbcbQ"
  color="#026cdf"
>
  snapshot test
</div>
`;
